File: component-acgschedule/src/main/java/com/rabtman/acgschedule/mvp/model/dao/ScheduleDAO.java
package com.rabtman.acgschedule.mvp.model.dao;

import android.util.Pair;
import com.rabtman.acgschedule.mvp.model.entity.ScheduleCache;
import com.rabtman.common.utils.RxRealmUtils;
import io.reactivex.Flowable;
import io.reactivex.FlowableEmitter;
import io.reactivex.functions.Consumer;
import io.realm.Realm;
import io.realm.Realm.Transaction;
import io.realm.RealmConfiguration;
import io.realm.RealmResults;
import java.util.List;

/**
 * @author Rabtman
 */
public class ScheduleDAO {

  private RealmConfiguration realmConfiguration;

  public ScheduleDAO(RealmConfiguration realmConfiguration) {
    this.realmConfiguration = realmConfiguration;
  }

  public Flowable<ScheduleCache> addScheduleCache(final ScheduleCache item) {
    return RxRealmUtils
        .flowableExec(realmConfiguration, new Consumer<Pair<FlowableEmitter, Realm>>() {
          @Override
          public void accept(final Pair<FlowableEmitter, Realm> pair) throws Exception {
            pair.second.executeTransaction(new Transaction() {
              @Override
              public void execute(Realm r) {
                pair.first.onNext(r.copyFromRealm(r.copyToRealmOrUpdate(item)));
                pair.first.onComplete();
              }
            });
          }
        });
  }

  public Flowable<Boolean> deleteScheduleCacheByUrl(final String url) {
    return RxRealmUtils
        .flowableExec(realmConfiguration, new Consumer<Pair<FlowableEmitter, Realm>>() {
          @Override
          public void accept(final Pair<FlowableEmitter, Realm> pair) throws Exception {
            pair.second.executeTransactionAsync(new Transaction() {
              @Override
              public void execute(Realm r) {
                final boolean isSuccess = r.where(ScheduleCache.class)
                    .equalTo("scheduleUrl", url)
                    .findAll()
                    .deleteAllFromRealm();
                pair.first.onNext(isSuccess);
                pair.first.onComplete();
              }
            });
          }
        });
  }

  public Flowable<ScheduleCache> getScheduleCacheByUrl(String url) {
    try (final Realm realm = Realm.getInstance(realmConfiguration)) {
      ScheduleCache queryResult = realm.where(ScheduleCache.class)
          .equalTo("scheduleUrl", url)
          .findFirst();
      return Flowable.just(
          queryResult == null ? new ScheduleCache("", -1) : realm.copyFromRealm(queryResult)
      );
    }
  }

  public Flowable<List<ScheduleCache>> getScheduleCollectCaches() {
    try (final Realm realm = Realm.getInstance(realmConfiguration)) {
      RealmResults<ScheduleCache> queryResult = realm.where(ScheduleCache.class)
          .equalTo("isCollect", true)
          .findAll();
      if (queryResult != null) {
        return Flowable.just(realm.copyFromRealm(queryResult));
      } else {
        return Flowable.empty();
      }
    }
  }
}

File: component-acgschedule/src/main/java/com/rabtman/acgschedule/mvp/model/dao/ScheduleDAO.kt
package com.rabtman.acgschedule.mvp.model.dao

import com.rabtman.acgschedule.mvp.model.entity.ScheduleCache
import com.rabtman.common.utils.RxRealmUtils
import io.reactivex.Flowable
import io.realm.Realm
import io.realm.RealmConfiguration

/**
 * @author Rabtman
 */
class ScheduleDAO(private val realmConfiguration: RealmConfiguration) {
    fun addScheduleCache(item: ScheduleCache): Flowable<ScheduleCache> {
        return RxRealmUtils
                .flowableExec(realmConfiguration) { pair ->
                    pair.second.executeTransaction { r ->
                        pair.first.onNext(r.copyFromRealm(r.copyToRealmOrUpdate(item)))
                        pair.first.onComplete()
                    }
                }
    }

    fun deleteScheduleCacheByUrl(url: String?): Flowable<Boolean> {
        return RxRealmUtils
                .flowableExec(realmConfiguration) { pair ->
                    pair.second.executeTransactionAsync { r ->
                        val isSuccess = r.where(ScheduleCache::class.java)
                                .equalTo("scheduleUrl", url)
                                .findAll()
                                .deleteAllFromRealm()
                        pair.first.onNext(isSuccess)
                        pair.first.onComplete()
                    }
                }
    }

    fun getScheduleCacheByUrl(url: String?): Flowable<ScheduleCache> {
        Realm.getInstance(realmConfiguration).use { realm ->
            val queryResult = realm.where(ScheduleCache::class.java)
                    .equalTo("scheduleUrl", url)
                    .findFirst()
            return Flowable.just(
                    if (queryResult == null) ScheduleCache("", -1) else realm.copyFromRealm(queryResult)
            )
        }
    }

    val scheduleCollectCaches: Flowable<List<ScheduleCache>>
        get() {
            Realm.getInstance(realmConfiguration).use { realm ->
                val queryResult = realm.where(ScheduleCache::class.java)
                        .equalTo("isCollect", true)
                        .findAll()
                return if (queryResult != null) {
                    Flowable.just(realm.copyFromRealm(queryResult))
                } else {
                    Flowable.empty()
                }
            }
        }

}